### PR TITLE
SONAR-7951 SONAR-7952 Duplication metrics on new code

### PR DESCRIPTION
**Pull request by teryk**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/NewDuplicationMeasuresStep.java [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+
+    private final DuplicationRepository duplicationRepository;
+    private final ScmInfoRepository scmInfoRepository;
+
+    private NewDuplicationFormula(ScmInfoRepository scmInfoRepository, @Nullable DuplicationRepository duplicationRepository) {
+      this.duplicationRepository = duplicationRepository;
+      this.scmInfoRepository = scmInfoRepository;
+    }
+
+    public static Formula<?> from(@Nullable ScmInfoRepository scmInfoRepository, @Nullable DuplicationRepository duplicationRepository) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/NewDuplicationMeasuresStep.java]: MAJOR Remove usage of generic wildcard type. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1452)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/NewDuplicationMeasuresStep.java [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+    ScmInfoRepository scmInfoRepository, @Nullable DuplicationRepository duplicationRepository) {
+    this.treeRootHolder = treeRootHolder;
+    this.periodsHolder = periodsHolder;
+    this.metricRepository = metricRepository;
+    this.measureRepository = measureRepository;
+    this.formulas = ImmutableList.of(NewDuplicationFormula.from(scmInfoRepository, duplicationRepository));
+  }
+
+  /**
+   * Constructor used by Pico in Governance where no DuplicationRepository is available.
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/NewDuplicationMeasuresStep.java]: You forget to say that this constructor is also used when no ScmInfoRepository is available

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/NewDuplicationMeasuresStep.java [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+
+  private final ImmutableList<Formula> formulas;
+
+  private final TreeRootHolder treeRootHolder;
+  private final PeriodsHolder periodsHolder;
+  private final MetricRepository metricRepository;
+  private final MeasureRepository measureRepository;
+
+  public NewDuplicationMeasuresStep(TreeRootHolder treeRootHolder, PeriodsHolder periodsHolder, MetricRepository metricRepository, MeasureRepository measureRepository,
+    ScmInfoRepository scmInfoRepository, @Nullable DuplicationRepository duplicationRepository) {
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/NewDuplicationMeasuresStep.java]: ScmInfoRepository should have the nullable annotation

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/NewDuplicationMeasuresStep.java [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+    this.periodsHolder = periodsHolder;
+    this.metricRepository = metricRepository;
+    this.measureRepository = measureRepository;
+    this.formulas = ImmutableList.of(NewDuplicationFormula.from(scmInfoRepository, duplicationRepository));
+  }
+
+  /**
+   * Constructor used by Pico in Governance where no DuplicationRepository is available.
+   */
+  public NewDuplicationMeasuresStep(TreeRootHolder treeRootHolder, PeriodsHolder periodsHolder, MetricRepository metricRepository, MeasureRepository measureRepository) {
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/NewDuplicationMeasuresStep.java]: No unit test is using this constructor

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/NewDuplicationMeasuresStep.java [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+      FormulaExecutorComponentVisitor.newBuilder(metricRepository, measureRepository)
+        .withVariationSupport(periodsHolder)
+        .buildFor(formulas))
+          .visit(treeRootHolder.getRoot());
+  }
+
+  private static class NewDuplicationCounter implements Counter<NewDuplicationCounter> {
+    @CheckForNull
+    private final DuplicationRepository duplicationRepository;
+    private final ScmInfoRepository scmInfoRepository;
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/NewDuplicationMeasuresStep.java]: Add missing CheckForNull annotation

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/NewDuplicationMeasuresStep.java (hunk 4 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/NewDuplicationMeasuresStep.java]: Forget it, it's used in views tests

Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/NewDuplicationMeasuresStep.java]: As discussed, this step should not handled views computation, it should be done in governance.

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/NewDuplicationMeasuresStep.java [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+
+    @Override
+    public void aggregate(NewDuplicationCounter counter) {
+      this.newLines.incrementAll(counter.newLines);
+    }
+
+    @Override
+    public void initialize(CounterInitializationContext context) {
+      Component leaf = context.getLeaf();
+      Iterable<Duplication> duplications = requireNonNull(this.duplicationRepository, "DuplicationRepository missing").getDuplications(leaf);
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/NewDuplicationMeasuresStep.java]: I don't see any use case when duplicationRepository could be null here, you should remove the requireNonNull

File: server/sonar-server/src/test/java/org/sonar/server/computation/task/projectanalysis/step/ReportNewDuplicationMeasuresStepTest.java [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+  public MetricRepositoryRule metricRepository = new MetricRepositoryRule()
+    .add(NEW_LINES_DUPLICATED);
+  @Rule
+  public MeasureRepositoryRule measureRepository = MeasureRepositoryRule.create(treeRootHolder, metricRepository);
+  @Rule
+  public DuplicationRepositoryRule duplicationRepository = DuplicationRepositoryRule.create(treeRootHolder);
+
+  @Before
+  public void setUp() {
+    periodsHolder.setPeriods(
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/computation/task/projectanalysis/step/ReportNewDuplicationMeasuresStepTest.java]: It's not needed to define periods in setup(), you can directly defined them when creating the PeriodsHolderRule (line 89).

File: server/sonar-server/src/test/java/org/sonar/server/computation/task/projectanalysis/step/ReportNewDuplicationMeasuresStepTest.java [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+      new Period(5, "mode_p_5", null, parseDate("2011-02-18").getTime(), "u2"));
+  }
+
+  NewDuplicationMeasuresStep underTest = new NewDuplicationMeasuresStep(treeRootHolder, periodsHolder, metricRepository, measureRepository, scmInfoRepository,
+    duplicationRepository);
+
+  @Test
+  public void compute_duplicated_lines_counts_lines_from_original_and_InnerDuplicate_of_a_single_line() {
+    TextBlock original = new TextBlock(1, 1);
+    duplicationRepository.addDuplication(FILE_1_REF, original, new TextBlock(2, 2));
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/computation/task/projectanalysis/step/ReportNewDuplicationMeasuresStepTest.java]: original should be replaced by new TextBlock(1, 1). No need to declare a variable for that.
(and do the same on following tests)

File: server/sonar-server/src/test/java/org/sonar/server/computation/task/projectanalysis/step/ReportNewDuplicationMeasuresStepTest.java [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+    assertRawMeasureValue(FILE_3_REF, 0);
+    assertRawMeasureValue(FILE_4_REF, 0);
+    assertRawMeasureValue(DIRECTORY_REF, 0);
+    assertRawMeasureValue(SUB_MODULE_1_REF, 0);
+    assertRawMeasureValue(SUB_MODULE_2_REF, 0);
+    assertRawMeasureValue(MODULE_REF, 0);
+    assertRawMeasureValue(ROOT_REF, 0);
+  }
+
+  private static Measure createMeasure(@Nullable Double variationPeriod2, @Nullable Double variationPeriod5) {
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/computation/task/projectanalysis/step/ReportNewDuplicationMeasuresStepTest.java]: Remove this unused method

**Comment by julienlancelot** [comment]
Please correctly handle computation of these measures on views.

**Comment by teryk** [comment]
as stated, computation for views are not handled in this MMF, tickets will be created to handle the metrics in Governance

**Comment by julienlancelot** [comment]
LGTM !

**Pull request merged**